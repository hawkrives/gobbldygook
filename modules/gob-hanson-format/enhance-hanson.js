// @flow
import isRequirementName from '@gob/examine-student/source/is-requirement-name'
import fromPairs from 'lodash/fromPairs'
import {makeAreaSlug} from './make-area-slug'
const {parse} = require('./parse-hanson-string')

const requirementNameRegex = /(.*?) +\(([A-Z-]+)\)$/i
const quote = str => `"${str}"`
const quoteAndJoin = list => [...list].map(quote).join(', ')

const baseWhitelist = ['result', 'message', 'declare', 'children share courses']
const topLevelWhitelist = new Set([
	...baseWhitelist,
	'name',
	'revision',
	'type',
	'sourcePath',
	'slug',
	'source',
	'dateAdded',
	'available through',
	'_error',
])
const lowerLevelWhitelist = new Set([
	...baseWhitelist,
	'filter',
	'message',
	'description',
	'student selected',
	'contract',
])

const startRules = {
	result: 'Result',
	filter: 'Filter',
}

type StringMap = {[key: string]: string}

type HansonData = {
	[key: string]: mixed,
	name?: string,
	revision?: string,
	declare?: StringMap,
	result: mixed,
}

export function enhanceHanson(
	data: HansonData,
	{
		topLevel = true,
		declaredVariables = {},
	}: {topLevel: boolean, declaredVariables: StringMap} = {},
) {
	// 1. adds 'result' key, if missing
	// 2. parses the 'result' and 'filter' keys
	// 3. throws if it encounters any lowercase keys not in the whitelist
	// 4. throws if it cannot find any of the required keys

	if (typeof data !== 'object') {
		throw new Error('enhanceHanson: data was not an object!')
	}

	// Ensure that a result, message, or filter key exists.
	// If filter's the only one, it's going to filter the list of courses
	// available to the child requirements when this is evaluated.
	const oneOfTheseKeysMustExist = new Set(['result', 'message', 'filter'])
	if (!Object.keys(data).some(key => oneOfTheseKeysMustExist.has(key))) {
		let requiredKeys = quoteAndJoin(oneOfTheseKeysMustExist)
		let existingKeys = quoteAndJoin(Object.keys(data))
		throw new TypeError(
			`enhanceHanson(): could not find any of [${requiredKeys}] in [${existingKeys}].`,
		)
	}

	const whitelist = topLevel ? topLevelWhitelist : lowerLevelWhitelist

	Object.keys(data).forEach(key => {
		if (!isRequirementName(key) && !whitelist.has(key)) {
			const whitelistStr = quoteAndJoin(whitelist)
			throw new TypeError(
				`enhanceHanson: only [${whitelistStr}] keys are allowed, and '${key}' is not one of them. All requirement names must begin with an uppercase letter or a number.`,
			)
		}
	})

	// install the top-level $type key *after* the whitelist runs
	if (topLevel) {
		data.$type = 'requirement'

		// because this only runs at the top level, we know
		// that we'll have a name to use
		data.slug = data.slug || makeAreaSlug(data.name || '')

		if (data.revision && typeof data.revision !== 'string') {
			let rev = data.revision
			throw new TypeError(
				`enhanceHanson: "revision" must be a string. Try wrapping it in single quotes. "${rev}" is a ${typeof rev}.`,
			)
		}
	}

	// Create the lists of requirement titles and abbreviations for the parser.
	// Because we allow both full titles ("Biblical Studies") and shorthand
	// abbreviations ("BTS-B") all glommed together into one string ("Biblical
	// Studies (BTS-B)"), we need a method of splitting those apart so the
	// PEG's ReferenceExpression can correctly reference them.
	const requirements = Object.keys(data).filter(isRequirementName)
	const abbreviations = fromPairs(
		requirements.map(req => [req.replace(requirementNameRegex, '$2'), req]),
	)
	const titles = fromPairs(
		requirements.map(req => [req.replace(requirementNameRegex, '$1'), req]),
	)

	// (Variables)
	// We load the list of variables with the keys listed in the `declare` key
	// into the declaredVariables map. They're defined as a [string: string]
	// mapping.
	declaredVariables = data.declare || {}

	const mutated = Object.entries(data).map(([key: string, value: mixed]) => {
		if (isRequirementName(key)) {
			// expand simple strings into {result: string} objects
			if (typeof value === 'string') {
				value = {result: value}
			}

			// then run enhance on the resultant object
			value = enhanceHanson((value: any), {
				topLevel: false,
				declaredVariables,
			})

			// also set $type; the PEG can't do it b/c the spec file is YAML
			// w/ PEG result strings.
			value.$type = 'requirement'

			return [key, value]
		}

		if (key === 'result' || key === 'filter') {
			if (typeof value !== 'string') {
				throw new Error(
					`value ${String(
						value,
					)} should be a string, not a ${typeof value}`,
				)
			}

			// (Variables)
			// Next up, we go through the list of variables and look for any
			// occurrences of the named variables in the value, prefixed with
			// a $. So, for instance, the variable defined as "math-level-3"
			// would be referenced via "$math-level-3".
			for (let [name, contents] of Object.entries(declaredVariables)) {
				if (typeof contents !== 'string') {
					throw new Error(
						`value ${String(
							contents,
						)} should be a string, not a ${typeof contents}`,
					)
				}

				let exprName = `$${name}`

				// istanbul ignore else
				if (value.includes(exprName)) {
					value = value.split(exprName).join(contents)
				}
			}

			try {
				value = parse(value, {
					abbreviations,
					titles,
					startRule: startRules[key],
				})

				return [key, value]
			} catch (e) {
				throw new SyntaxError(
					`enhanceHanson: ${e.message} (in '${value}')`,
				)
			}
		}

		return [key, value]
	})

	return fromPairs(mutated)
}
