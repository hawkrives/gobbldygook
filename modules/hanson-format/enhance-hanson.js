// @flow
import isRequirementName from 'modules/core/examine-student/is-requirement-name'
import filter from 'lodash/filter'
import forEach from 'lodash/forEach'
import includes from 'lodash/includes'
import isString from 'lodash/isString'
import keys from 'lodash/keys'
import map from 'lodash/map'
import mapValues from 'lodash/mapValues'
import some from 'lodash/some'
import fromPairs from 'lodash/fromPairs'
import { makeAreaSlug } from './make-area-slug'
import { parse } from './parse-hanson-string'

const requirementNameRegex = /(.*?) +\(([A-Z\-]+)\)$/i
const none = (arr, pred) => !some(arr, pred)
const quote = str => `"${str}"`
const quoteAndJoin = list => list.map(quote).join(', ')

const baseWhitelist = [ 'result', 'message', 'declare', 'children share courses' ]
const topLevelWhitelist = baseWhitelist.concat([ 'name', 'revision', 'type', 'sourcePath', 'slug', 'source', 'dateAdded', 'available through', '_error' ])
const lowerLevelWhitelist = baseWhitelist.concat([ 'filter', 'message', 'description', 'student selected' ])

const startRules = {
	'result': 'Result',
	'filter': 'Filter',
}

type StringMap = {[key: string]: string};

export function enhanceHanson(data: any, { topLevel=true, declaredVariables={} }: {topLevel: boolean, declaredVariables: StringMap}={}) {
	// 1. adds 'result' key, if missing
	// 2. parses the 'result' and 'filter' keys
	// 3. throws if it encounters any lowercase keys not in the whitelist
	// 4. throws if it cannot find any of the required keys

	if (typeof data !== 'object') {
		throw new Error('enhanceHanson: data was not an object!')
	}

	// Ensure that a result, message, or filter key exists.
	// If filter's the only one, it's going to filter the list of courses
	// available to the child requirements when this is evaluated.
	const oneOfTheseKeysMustExist = [ 'result', 'message', 'filter' ]
	if (none(keys(data), key => includes(oneOfTheseKeysMustExist, key))) {
		let requiredKeys = quoteAndJoin(oneOfTheseKeysMustExist)
		let existingKeys = quoteAndJoin(keys(data))
		throw new TypeError(`enhanceHanson(): could not find any of [${requiredKeys}] in [${existingKeys}].`)
	}

	const whitelist = topLevel ? topLevelWhitelist : lowerLevelWhitelist

	forEach(keys(data), key => {
		if (!isRequirementName(key) && !includes(whitelist, key)) {
			throw new TypeError(`enhanceHanson: only [${quoteAndJoin(whitelist)}] keys are allowed, and '${key}' is not one of them. All requirement names must begin with an uppercase letter or a number.`)
		}
	})

	// install the top-level $type key *after* the whitelist runs
	if (topLevel) {
		data.$type = 'requirement'

		// because this only runs at the top level, we know
		// that we'll have a name to use
		data.slug = data.slug || makeAreaSlug(data.name)

		if (data.revision && typeof data.revision !== 'string') {
			throw new TypeError(`enhanceHanson: "revision" must be a string. Try wrapping it in single quotes. "${data.revision}" is a ${typeof data.revision}.`)
		}
	}

	// Create the lists of requirement titles and abbreviations for the parser.
	// Because we allow both full titles ("Biblical Studies") and shorthand
	// abbreviations ("BTS-B") all glommed together into one string ("Biblical
	// Studies (BTS-B)"), we need a method of splitting those apart so the
	// PEG's ReferenceExpression can correctly reference them.
	const requirements = filter(keys(data), isRequirementName)
	const abbreviations = fromPairs(map(requirements, req => [ req.replace(requirementNameRegex, '$2'), req ]))
	const titles = fromPairs(map(requirements, req => [ req.replace(requirementNameRegex, '$1'), req ]))

	// (Variables)
	// We load the list of variables with the keys listed in the `declare` key
	// into the declaredVariables map. They're defined as a [string: string]
	// mapping.
	declaredVariables = data.declare || {}

	const mutated = mapValues(data, (value, key) => {
		if (isRequirementName(key)) {
			// expand simple strings into {result: string} objects
			if (isString(value)) {
				value = { result: value }
			}

			// then run enhance on the resultant object
			value = enhanceHanson(value, { topLevel: false, declaredVariables })

			// also set $type; the PEG can't do it b/c the spec file is YAML
			// w/ PEG result strings.
			value.$type = 'requirement'
		}

		else if (key === 'result' || key === 'filter') {
			if (typeof value !== 'string') {
				throw new Error(`value ${value.toString()} should be a string, not a ${typeof value}`)
			}

			// (Variables)
			// Next up, we go through the list of variables and look for any
			// occurrences of the named variables in the value, prefixed with
			// a $. So, for instance, the variable defined as "math-level-3"
			// would be referenced via "$math-level-3".
			forEach(declaredVariables, (contents, name) => {
				// _we_ know value is a string here
				let val = ((value: any): string)
				// istanbul ignore else
				if (includes(val, '$' + name)) {
					value = val.split(`$${name}`).join(contents)
				}
			})

			try {
				value = parse(value, { abbreviations, titles, startRule: startRules[key] })
			}
			catch (e) {
				throw new SyntaxError(`enhanceHanson: ${e.message} (in '${value}')`)
			}
		}

		return value
	})

	return mutated
}
