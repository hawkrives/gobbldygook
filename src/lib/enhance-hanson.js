import {parse} from './parse-hanson-string'
import cloneDeep from 'lodash/lang/cloneDeep'
import filter from 'lodash/collection/filter'
import forEach from 'lodash/collection/forEach'
import {oxford} from 'humanize-plus'
import includes from 'lodash/collection/includes'
import isRequirementName from './is-requirement-name'
import isString from 'lodash/lang/isString'
import keys from 'lodash/object/keys'
import map from 'lodash/collection/map'
import mapValues from 'lodash/object/mapValues'
import zipObject from 'lodash/array/zipObject'

let declaredVariables = {}

export default function enhanceHanson(data, {topLevel=false}={}) {
	// 1. adds 'result' key, if missing
	// 2. parses the 'result' and 'filter' keys
	// 3. warns if it encounters any lowercase keys not in the whitelist

	const baseWhitelist = ['result', 'message', 'declare']
	const topLevelWhitelist = baseWhitelist.concat(['name', 'revision', 'type', 'sourcePath', '_id', '_rev'])
	const lowerLevelWhitelist = baseWhitelist.concat(['filter', 'message', 'description'])
	const whitelist = topLevel ? topLevelWhitelist : lowerLevelWhitelist

	forEach(keys(data), key => {
		if (!isRequirementName(key) && !includes(whitelist, key)) {
			throw new TypeError(`enhanceHanson(): only ${oxford(whitelist)} keys are allowed, and '${key}' is not one of them. all requirements must begin with an uppercase letter or a number.`)
		}
	})

	// install the top-level $type key *after* the whitelist runs
	if (topLevel) {
		data.$type = 'requirement'
	}

	const requirements = filter(keys(data), isRequirementName)
	const abbreviations = zipObject(map(requirements,
		req => [req.replace(/.* \(([A-Z\-]+)\)$|.*$/, '$1'), req]))
	const titles = zipObject(map(requirements,
		req => [req.replace(/(.*?) +\([A-Z\-]+\)$|.*$/, '$1'), req]))

	// console.log(abbreviations, titles)

	const oldVariables = cloneDeep(declaredVariables)
	declaredVariables = {}

	forEach(data.declare || [], (value, key) => {
		declaredVariables[key] = value
	})

	const mutated = mapValues(data, (value, key) => {
		if (isRequirementName(key)) {
			// expand simple strings into {result: string} objects
			if (isString(value)) {
				value = {result: value}
			}

			// then run enhance on the resultant object
			value = enhanceHanson(value, {topLevel: false})

			// also set $type; the PEG can't do it b/c the spec file is YAML
			// w/ PEG result strings.
			value.$type = 'requirement'
		}

		else if (key === 'result' || key === 'filter') {
			forEach(declaredVariables, (contents, name) => {
				if (includes(value, '$' + name)) {
					value = value.split(`$${name}`).join(contents)
				}
			})

			try {
				value = parse(value, {abbreviations, titles})
			}
			catch (e) {
				throw new SyntaxError(`enhanceHanson(): ${e.message} (in '${value}')`)
			}
		}

		return value
	})

	forEach(data.declare || [], (value, key) => {
		delete declaredVariables[key]
	})
	declaredVariables = oldVariables

	return mutated
}
