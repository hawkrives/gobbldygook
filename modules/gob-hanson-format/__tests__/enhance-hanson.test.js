// @flow

import {enhanceHanson, enhanceHansonRequirement} from '../enhance-hanson'

const basicFile = () => ({
	name: 'test',
	type: 'test',
	revision: 'test',
	requirements: {},
	result: 'CSCI 121',
})

describe('enhanceHanson', () => {
	it('adds a "slug" key to the top-level', () => {
		const actual = enhanceHanson({
			name: 'test',
			type: 'test',
			revision: 'test',
			requirements: {},
			result: 'CSCI 121',
		})

		expect(actual).toMatchInlineSnapshot(`
Object {
  "name": "test",
  "requirements": Object {},
  "result": Object {
    "$type": "course",
    "department": "CSCI",
    "number": 121,
  },
  "revision": "test",
  "slug": "test",
  "type": "test",
}
`)
		expect(actual.slug).toBe('test')
	})

	it('requires the top-level to have certain keys', () => {
		// $FlowExpectedError
		expect(() => enhanceHanson({slug: 'nope'})).toThrowError(
			'enhanceHanson(): "name" is missing in ["slug"] (need all of ["name", "requirements", "result", "revision", "type"]).',
		)

		// $FlowExpectedError
		expect(() => enhanceHanson({name: 'have a nice day'})).toThrowError(
			'enhanceHanson(): "requirements" is missing in ["name"] (need all of ["name", "requirements", "result", "revision", "type"]).',
		)

		expect(() => enhanceHanson({...basicFile()})).not.toThrow()
	})

	it('requires its input to be an object', () => {
		// $FlowExpectedError
		expect(() => enhanceHanson('')).toThrowError(
			'enhanceHanson: data was not an object!',
		)
	})

	it('requires "revision" to be a string, if present', () => {
		expect(() =>
			// $FlowExpectedError
			enhanceHanson({...basicFile(), revision: 2}),
		).toThrowError(
			'enhanceHanson: "revision" must be a string. Try wrapping it in single quotes.',
		)

		expect(() =>
			enhanceHanson({...basicFile(), revision: '2'}),
		).not.toThrow()
	})

	it('enforces a whitelist of keys at the top-level', () => {
		expect(() => enhanceHanson({...basicFile(), xxx: 'yyy'})).toThrowError(
			/only \[.*\] keys are allowed/,
		)
	})

	it('assumes that keys starting with a capital letter are requirements', () => {
		expect(() =>
			enhanceHanson({
				...basicFile(),
				result: 'Req',
				requirements: {
					Req: 'CSCI 121',
				},
			}),
		).not.toThrow()
	})

	it('enforces a whitelist of keys at lower levels', () => {
		expect(() =>
			enhanceHansonRequirement({result: '', innerbad: 'zzzz'}),
		).toThrow(/only \[.*\] keys are allowed/)
	})

	it('expands string-only keys into objects with a "result" key', () => {
		const actual = enhanceHanson({
			...basicFile(),
			result: 'Requirement',
			requirements: {
				Requirement: 'CSCI 121',
			},
		})
		expect(actual).toMatchInlineSnapshot(`
Object {
  "name": "test",
  "requirements": Object {
    "Requirement": Object {
      "$type": "requirement",
      "result": Object {
        "$type": "course",
        "department": "CSCI",
        "number": 121,
      },
    },
  },
  "result": Object {
    "$requirement": "Requirement",
    "$type": "reference",
  },
  "revision": "test",
  "slug": "test",
  "type": "test",
}
`)
	})

	it('parses "result" strings with the Result PEG rule', () => {
		expect(() =>
			enhanceHanson({
				...basicFile(),
				result: 'Req',
				requirements: {
					Req: {
						result: 'only courses from (CSCI 121)',
					},
				},
			}),
		).toThrowError('enhanceHanson: Expected expression but "o" found.')

		expect(() =>
			enhanceHanson({
				...basicFile(),
				result: 'Req',
				requirements: {
					Req: {
						result: 'one of (CSCI 121)',
					},
				},
			}),
		).not.toThrow()

		expect(
			enhanceHanson({
				...basicFile(),
				result: 'Req',
				requirements: {
					Req: {
						result: 'one of (CSCI 121)',
					},
				},
			}),
		).toMatchInlineSnapshot(`
Object {
  "name": "test",
  "requirements": Object {
    "Req": Object {
      "$type": "requirement",
      "result": Object {
        "$count": Object {
          "$num": 1,
          "$operator": "$gte",
        },
        "$of": Array [
          Object {
            "$type": "course",
            "department": "CSCI",
            "number": 121,
          },
        ],
        "$type": "of",
      },
    },
  },
  "result": Object {
    "$requirement": "Req",
    "$type": "reference",
  },
  "revision": "test",
  "slug": "test",
  "type": "test",
}
`)
	})

	it('parses "filter" strings with the Filter PEG rule', () => {
		expect(() =>
			enhanceHanson({
				...basicFile(),
				result: 'Req',
				requirements: {
					Req: {
						filter: 'only courses from (CSCI 121)',
						result: 'any of (Req)',
					},
				},
			}),
		).not.toThrow()

		expect(
			enhanceHanson({
				...basicFile(),
				result: 'Req',
				requirements: {
					Req: {
						filter: 'only courses from (CSCI 121)',
						result: 'any of (Req)',
					},
				},
			}),
		).toMatchInlineSnapshot(`
Object {
  "name": "test",
  "requirements": Object {
    "Req": Object {
      "$type": "requirement",
      "filter": Object {
        "$distinct": false,
        "$filterType": "of",
        "$of": Array [
          Object {
            "$type": "course",
            "department": "CSCI",
            "number": 121,
          },
        ],
        "$type": "filter",
      },
      "result": Object {
        "$count": Object {
          "$num": 1,
          "$operator": "$gte",
          "$was": "any",
        },
        "$of": Array [
          Object {
            "$requirement": "Req",
            "$type": "reference",
          },
        ],
        "$type": "of",
      },
    },
  },
  "result": Object {
    "$requirement": "Req",
    "$type": "reference",
  },
  "revision": "test",
  "slug": "test",
  "type": "test",
}
`)

		expect(() =>
			enhanceHanson({
				...basicFile(),
				result: 'Req',
				requirements: {
					Req: {
						filter: 'one of (CSCI 121)',
						result: 'Req',
					},
				},
			}),
		).toThrowError('enhanceHanson: Expected "only" but "o" found.')
	})

	it('allows defining variables in result', () => {
		const input = {
			...basicFile(),
			result: 'Req',
			requirements: {
				Req: {
					declare: {
						'math-level-3': 'MATH 330, 340, 344',
					},
					result: 'one of ($math-level-3)',
				},
			},
		}

		const output = enhanceHanson(input)

		// $FlowExpectedError
		expect(output.requirements.Req.result.$of.length).toBe(3)

		expect(output).toMatchInlineSnapshot(`
Object {
  "name": "test",
  "requirements": Object {
    "Req": Object {
      "$type": "requirement",
      "result": Object {
        "$count": Object {
          "$num": 1,
          "$operator": "$gte",
        },
        "$of": Array [
          Object {
            "$type": "course",
            "department": "MATH",
            "number": 330,
          },
          Object {
            "$type": "course",
            "department": "MATH",
            "number": 340,
          },
          Object {
            "$type": "course",
            "department": "MATH",
            "number": 344,
          },
        ],
        "$type": "of",
      },
    },
  },
  "result": Object {
    "$requirement": "Req",
    "$type": "reference",
  },
  "revision": "test",
  "slug": "test",
  "type": "test",
}
`)
	})

	it('allows using variables in the filter', () => {
		const input = {
			...basicFile(),
			result: 'Req',
			requirements: {
				Req: {
					declare: {
						'math-level-3': 'MATH 330, 340, 344',
					},
					filter: 'only courses from ($math-level-3)',
					result: 'Req',
				},
			},
		}

		const output = enhanceHanson(input)

		// $FlowExpectedError
		expect(output.requirements.Req.filter.$of.length).toBe(3)
		expect(output).toMatchInlineSnapshot(`
Object {
  "name": "test",
  "requirements": Object {
    "Req": Object {
      "$type": "requirement",
      "filter": Object {
        "$distinct": false,
        "$filterType": "of",
        "$of": Array [
          Object {
            "$type": "course",
            "department": "MATH",
            "number": 330,
          },
          Object {
            "$type": "course",
            "department": "MATH",
            "number": 340,
          },
          Object {
            "$type": "course",
            "department": "MATH",
            "number": 344,
          },
        ],
        "$type": "filter",
      },
      "result": Object {
        "$requirement": "Req",
        "$type": "reference",
      },
    },
  },
  "result": Object {
    "$requirement": "Req",
    "$type": "reference",
  },
  "revision": "test",
  "slug": "test",
  "type": "test",
}
`)
	})

	it('only allows the variables to be used where they are declared', () => {
		const input = {
			...basicFile(),
			result: 'Parent',
			requirements: {
				Parent: {
					declare: {
						'math-level-3': 'MATH 330, 340, 344',
					},
					Req: {
						result: 'one of ($math-level-3)',
					},
					result: 'Req',
				},
			},
		}

		expect(() => enhanceHanson(input)).toThrowError(
			'enhanceHanson: Expected expression but "o" found. (in \'one of ($math-level-3)\')',
		)
	})
})
