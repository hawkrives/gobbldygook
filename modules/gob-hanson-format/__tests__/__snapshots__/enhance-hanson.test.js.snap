// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enhanceHanson allows defining variables in result 1`] = `
Object {
  "$type": "requirement",
  "Req": Object {
    "$type": "requirement",
    "result": Object {
      "$count": Object {
        "$num": 1,
        "$operator": "$gte",
      },
      "$of": Array [
        Object {
          "$course": Object {
            "department": "MATH",
            "number": 330,
          },
          "$type": "course",
        },
        Object {
          "$course": Object {
            "department": "MATH",
            "number": 340,
          },
          "$type": "course",
        },
        Object {
          "$course": Object {
            "department": "MATH",
            "number": 344,
          },
          "$type": "course",
        },
        Object {
          "$course": Object {
            "department": "MATH",
            "number": 348,
          },
          "$type": "course",
        },
        Object {
          "$course": Object {
            "department": "MATH",
            "number": 351,
          },
          "$type": "course",
        },
        Object {
          "$course": Object {
            "department": "MATH",
            "number": 356,
          },
          "$type": "course",
        },
        Object {
          "$course": Object {
            "department": "MATH",
            "number": 364,
          },
          "$type": "course",
        },
        Object {
          "$course": Object {
            "department": "MATH",
            "number": 382,
          },
          "$type": "course",
        },
        Object {
          "$course": Object {
            "department": "MATH",
            "number": 384,
          },
          "$type": "course",
        },
      ],
      "$type": "of",
    },
  },
  "name": undefined,
  "result": Object {
    "$requirement": "Req",
    "$type": "reference",
  },
  "revision": undefined,
  "slug": "",
  "type": undefined,
}
`;

exports[`enhanceHanson allows using variables in the filter 1`] = `
Object {
  "$type": "requirement",
  "Req": Object {
    "$type": "requirement",
    "filter": Object {
      "$distinct": false,
      "$filterType": "of",
      "$of": Array [
        Object {
          "$course": Object {
            "department": "MATH",
            "number": 330,
          },
          "$type": "course",
        },
        Object {
          "$course": Object {
            "department": "MATH",
            "number": 340,
          },
          "$type": "course",
        },
        Object {
          "$course": Object {
            "department": "MATH",
            "number": 344,
          },
          "$type": "course",
        },
        Object {
          "$course": Object {
            "department": "MATH",
            "number": 348,
          },
          "$type": "course",
        },
        Object {
          "$course": Object {
            "department": "MATH",
            "number": 351,
          },
          "$type": "course",
        },
        Object {
          "$course": Object {
            "department": "MATH",
            "number": 356,
          },
          "$type": "course",
        },
        Object {
          "$course": Object {
            "department": "MATH",
            "number": 364,
          },
          "$type": "course",
        },
        Object {
          "$course": Object {
            "department": "MATH",
            "number": 382,
          },
          "$type": "course",
        },
        Object {
          "$course": Object {
            "department": "MATH",
            "number": 384,
          },
          "$type": "course",
        },
      ],
      "$type": "filter",
    },
  },
  "name": undefined,
  "result": Object {
    "$requirement": "Req",
    "$type": "reference",
  },
  "revision": undefined,
  "slug": "",
  "type": undefined,
}
`;

exports[`enhanceHanson expands string-only keys into objects with a "result" key 1`] = `
Object {
  "$type": "requirement",
  "Requirement": Object {
    "$type": "requirement",
    "result": Object {
      "$course": Object {
        "department": "CSCI",
        "number": 121,
      },
      "$type": "course",
    },
  },
  "name": undefined,
  "result": Object {
    "$requirement": "Requirement",
    "$type": "reference",
  },
  "revision": undefined,
  "slug": "",
  "type": undefined,
}
`;

exports[`enhanceHanson parses "filter" strings with the Filter PEG rule 1`] = `
Object {
  "$type": "requirement",
  "Req": Object {
    "$type": "requirement",
    "filter": Object {
      "$distinct": false,
      "$filterType": "of",
      "$of": Array [
        Object {
          "$course": Object {
            "department": "CSCI",
            "number": 121,
          },
          "$type": "course",
        },
      ],
      "$type": "filter",
    },
  },
  "name": undefined,
  "result": Object {
    "$requirement": "Req",
    "$type": "reference",
  },
  "revision": undefined,
  "slug": "",
  "type": undefined,
}
`;

exports[`enhanceHanson parses "result" strings with the Result PEG rule 1`] = `
Object {
  "$type": "requirement",
  "Req": Object {
    "$type": "requirement",
    "result": Object {
      "$count": Object {
        "$num": 1,
        "$operator": "$gte",
      },
      "$of": Array [
        Object {
          "$course": Object {
            "department": "CSCI",
            "number": 121,
          },
          "$type": "course",
        },
      ],
      "$type": "of",
    },
  },
  "name": undefined,
  "result": Object {
    "$requirement": "Req",
    "$type": "reference",
  },
  "revision": undefined,
  "slug": "",
  "type": undefined,
}
`;
